sources/obstacles/rectangular_obstacle.cpp

```diff
@@ -0,0 +1,223 @@
+    }
+
+    bool RectangularObstacle::isColliding(const RectangularObstacle &obs) const
+    {
+        if (rotation_center_.squaredDistance(obs.rotation_center_) >=
+            (half_diagonal_ + obs.half_diagonal_) * (half_diagonal_ + obs.half_diagonal_))
+            return false;
+
+        return !test_separation(left_bottom_corner_.getX(), right_bottom_corner_.getX(),
+                                getXToObstacleCoordinateSystem(obs.left_bottom_corner_rotate_),
+                                getXToObstacleCoordinateSystem(obs.left_upper_corner_rotate_),
+                                getXToObstacleCoordinateSystem(obs.right_bottom_corner_rotate_),
+                                getXToObstacleCoordinateSystem(obs.right_upper_corner_rotate_))
+               && !test_separation(left_bottom_corner_.getY(), left_upper_corner_.getY(),
+                                   getYToObstacleCoordinateSystem(obs.left_bottom_corner_rotate_),
+                                   getYToObstacleCoordinateSystem(obs.left_upper_corner_rotate_),
+                                   getYToObstacleCoordinateSystem(obs.right_bottom_corner_rotate_),
+                                   getYToObstacleCoordinateSystem(obs.right_upper_corner_rotate_))
+               && !test_separation(obs.left_bottom_corner_.getX(), obs.right_bottom_corner_.getX(),
+                                   obs.getXToObstacleCoordinateSystem(left_bottom_corner_rotate_),
+                                   obs.getXToObstacleCoordinateSystem(left_upper_corner_rotate_),
+                                   obs.getXToObstacleCoordinateSystem(right_bottom_corner_rotate_),
+                                   obs.getXToObstacleCoordinateSystem(right_upper_corner_rotate_))
+               && !test_separation(obs.left_bottom_corner_.getY(), obs.left_upper_corner_.getY(),
+                                   obs.getYToObstacleCoordinateSystem(left_bottom_corner_rotate_),
+                                   obs.getYToObstacleCoordinateSystem(left_upper_corner_rotate_),
+                                   obs.getYToObstacleCoordinateSystem(right_bottom_corner_rotate_),
+                                   obs.getYToObstacleCoordinateSystem(right_upper_corner_rotate_));
+    }
+
+    float RectangularObstacle::getHalfDiagonal() const
+    {
+        return half_diagonal_;
+    }
+
+    bool RectangularObstacle::operator==(const Obstacle &rhs) const
+    {
+        if (!Obstacle::operator==(rhs))
+            return false;
+
+        if (typeid(*this) != typeid(rhs))
+            return false;
+
+        RectangularObstacle ro_rhs = static_cast<const RectangularObstacle &>(rhs);
+        return angle_ == ro_rhs.angle_ && left_bottom_corner_ == ro_rhs.left_bottom_corner_ &&
+               right_upper_corner_ == ro_rhs.right_upper_corner_;
+    }
+
+    bool RectangularObstacle::test_separation(const float &a, const float &b, const float &a2, const float &b2,
```

Comment:
`RectangularObstacle::test_separation` can have the noexcept and constexpr specifiers

Reply:
std::min and std::max are not declared with noexcept... So i don't know if it's really a good idea.
In C++11: " constexpr functions may contain no more than a single executable statement: a return" Effective Modern C++ Scott Meyers, p99, but, i'm fine with C++14 ;-)